worktree_commit.go

```diff
@@ -0,0 +1,243 @@
+package git
+
+import (
+	"io"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/filemode"
+	"gopkg.in/src-d/go-git.v4/plumbing/format/index"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/go-git.v4/storage"
+	"gopkg.in/src-d/go-git.v4/utils/ioutil"
+
+	"gopkg.in/src-d/go-billy.v2"
+)
+
+// Commit stores the current contents of the index in a new commit along with
+// a log message from the user describing the changes.
+func (w *Worktree) Commit(msg string, opts *CommitOptions) (plumbing.Hash, error) {
```

Comment:
What you commit is what is in staging, is a very basic concept of git. If you want to do dirty things like storing an espontaneous commit just push is to the Storage. 

Since you are committing the staging and this is a concept of the worktree, I don't see why this should be at Repository.

I think that this is very natural way of doing a commit:
```go
r, err := Init(storage, fs)
w, err := r.Worktree()
billy.WriteFile(fs, "foo", []byte("foo"), 0644)
 w.Add("foo")
 w.Commit("foo\n", &CommitOptions{Author: defaultSignature()})
```